README.md
Signingscript
==============

[![Build Status](https://travis-ci.org/mozilla-releng/signingscript.svg?branch=master)](https://travis-ci.org/mozilla-releng/signingscript) [![Coverage Status](https://coveralls.io/repos/github/mozilla-releng/signingscript/badge.svg?branch=master)](https://coveralls.io/github/mozilla-releng/signingscript?branch=master)

This is designed to be run from scriptworker, but runs perfectly fine as a standalone script.


Testing
-------

Testing takes a few steps to set up.  Here's how:

### docker-signing-server

To test, you will need to point at a signing server.  Since production signing servers have restricted access and sensitive keys, it's easiest to point at a docker-signing-server instance locally during development.

To do so:

    git clone https://github.com/escapewindow/docker-signing-server
    cd docker-signing-server
    # Follow ./README.md to set up and run the docker instance

Remember the path to `./fake_ca/ca.crt` ; this will be the file that signingscript will use to verify the SSL connection.

### virtualenv

First, you need `python>=3.6.0`.

Next, create a python35 virtualenv, and install signingscript:

    # create the virtualenv in ./venv3
    virtualenv3 venv3
    # activate it
    . venv3/bin/activate
    # install signingscript from pypi
    pip install signingscript

If you want to use local clones of [signingscript](https://github.com/mozilla-releng/signingscript), [signtool](https://github.com/mozilla-releng/signtool), and/or [scriptworker](https://github.com/mozilla-releng/scriptworker), you can

    python setup.py develop

in each of the applicable directories after, or instead of the `pip install` command.

### password json

You'll need a password json file.  The format is

    {
      "BASE_CERT_SCOPE:dep-signing": [
        ["IPADDRESS:PORT", "USERNAME", "PASSWORD", ["SIGNING_FORMAT1", "SIGNING_FORMAT2"...]],
        ["SECOND_IPADDRESS:PORT", "USERNAME", "PASSWORD", ["SIGNING_FORMAT1", "SIGNING_FORMAT2"...]],
        ...
      ],
      "BASE_CERT_SCOPE:nightly-signing": [
        ["IPADDRESS:PORT", "USERNAME", "PASSWORD", ["SIGNING_FORMAT1", "SIGNING_FORMAT2"...]],
        ["SECOND_IPADDRESS:PORT", "USERNAME", "PASSWORD", ["SIGNING_FORMAT1", "SIGNING_FORMAT2"...]],
        ...
      ],
      "BASE_CERT_SCOPE:release-signing": [
        ["IPADDRESS:PORT", "USERNAME", "PASSWORD", ["SIGNING_FORMAT1", "SIGNING_FORMAT2"...]],
        ["SECOND_IPADDRESS:PORT", "USERNAME", "PASSWORD", ["SIGNING_FORMAT1", "SIGNING_FORMAT2"...]],
        ...
      ]
    }

This stripped down version will work with docker-signing-server:

    {
      "project:releng:signing:cert:dep-signing": [
        ["127.0.0.1:9110", "user", "pass", ["gpg"]]
      ]
    }

The user/pass for the docker-signing-server are `user` and `pass` for super sekrit security.

### config json

The config json looks like this (comments are not valid json, but I'm inserting comments for clarity.  Don't include the comments in the file!):


    {
      // path to the password json you created above
      "signing_server_config": "/src/signing/signingscript/example_server_config.json",

      // the work directory path.  task.json will live here, as well as downloaded binaries
      // this should be an absolute path.
      "work_dir": "/src/signing/work_dir",

      // the artifact directory path.  the signed binaries will be copied here for scriptworker to upload
      // this should be an absolute path.
      "artifact_dir": "/src/signing/artifact_dir",

      // the IP that docker-signing-server thinks you're coming from.
      // I got this value from running `docker network inspect bridge` and using the gateway.
      "my_ip": "172.17.0.1",

      // how many seconds should the signing token be valid for?
      "token_duration_seconds": 1200,

      // the path to the docker-signing-server fake_ca cert that you generated above.
      "ssl_cert": "/src/signing/docker-signing-server/fake_ca/ca.crt",

      // the path to signtool in your virtualenv that you created above
      "signtool": "/src/signing/venv3/bin/signtool",

      // enable debug logging
      "verbose": true,

      // the path to zipalign. This executable is usually present in $ANDROID_SDK_LOCATION/build-tools/$ANDROID_VERSION/zipalign
      "zipalign": "/absolute/path/to/zipalign",

      // the path to zip. This executable is usually present in /usr/bin/zip
      "zip": "/absolute/path/to/zip",

      // The host and port to use when sending metrics to datadog statsd
      // Datadog's default is 8125, ours is 8135 due to a conflict with collectd
      "datadog_port": 8135,
      "datadog_host": "localhost"

    }

#### directories and file naming
If you aren't running through scriptworker, you need to manually create the directories that `work_dir` and `artifact_dir` point to.  It's better to use new directories for these rather than cluttering and potentially overwriting an existing directory.  Once you set up scriptworker, the `work_dir` and `artifact_dir` will be regularly wiped and recreated.

Scriptworker will expect to find a config.json for the scriptworker config, so I name the signingscript config json `script_config.json`.  You can name it whatever you'd like.

### file to sign

Put the file(s) to sign somewhere where they can be reached via the web; you'll point to their URL(s) in the task.json below.  Alternately, point to the artifacts of a TaskCluster task, and add the `taskId` to your `dependencies` in the task.json below.

### task.json

Ordinarily, scriptworker would get the task definition from TaskCluster, and write it to a `task.json` in the `work_dir`.  Since you're initially not going to run through scriptworker, you need to put this file on disk yourself.

It will look like this:

    {
      "created": "2016-05-04T23:15:17.908Z",
      "deadline": "2016-05-05T00:15:17.908Z",
      "dependencies": [
        "VALID_TASK_ID"
      ],
      "expires": "2017-05-05T00:15:17.908Z",
      "extra": {},
      "metadata": {
        "description": "Markdown description of **what** this task does",
        "name": "Example Task",
        "owner": "name@example.com",
        "source": "https://tools.taskcluster.net/task-creator/"
      },
      "payload": {
        "upstreamArtifacts": [{
          "taskId": "upstream-task-id1",
          "taskType": "build",
          "paths": ["public/artifact/path1", "public/artifact/path2"],
          "formats": []
        }],
        "maxRunTime": 600
      },
      "priority": "normal",
      "provisionerId": "test-dummy-provisioner",
      "requires": "all-completed",
      "retries": 0,
      "routes": [],
      "schedulerId": "-",
      "scopes": [
        "project:releng:signing:cert:dep-signing",
        "project:releng:signing:format:gpg"
      ],
      "tags": {},
      "taskGroupId": "CRzxWtujTYa2hOs20evVCA",
      "workerType": "dummy-worker-aki"
    }

The important entries to edit are the `upstreamArtifacts`, the `dependencies`, and the `scopes`.

The `upstreamArtifacts` point to the file(s) to sign.  Because scriptworker downloads and verifies their shas, signingscript expects to find the files under `$work_dir/cot/$upstream_task_id/$path`

The first scope, `project:releng:signing:cert:dep-signing`, matches the scope in your password json that you created.  The second scope, `project:releng:signing:format:gpg`, specifies which signing format to use.  (You can specify multiple formats by adding multiple `project:releng:signing:format:` scopes)

Write this to `task.json` in your `work_dir`.

### run

You're ready to run signingscript!

    signingscript CONFIG_FILE

where `CONFIG_FILE` is the config json you created above.

This should download the file(s) specified in the payload, download a token from the docker-signing-server, upload the file(s) to the docker-signing-server to sign, download the signed bits from the docker-signing-server, and then copy the signed bits into the `artifact_dir`.

### troubleshooting

Invalid json is a common error.  Validate your json with this command:

    python -mjson.tool JSON_FILE

Your docker-signing-server shell should be able to read the `signing.log`, which should help troubleshoot.

### running through scriptworker

[Scriptworker](https://github.com/mozilla-releng/scriptworker) can deal with the TaskCluster specific parts, and run signingscript.

Follow the [scriptworker readme](https://github.com/mozilla-releng/scriptworker/blob/master/README.rst) to set up scriptworker, and use `["path/to/signingscript", "path/to/script_config.json"]` as your `task_script`.

Make sure your `work_dir` and `artifact_dir` point to the same directories between the scriptworker config and the signingscript config!


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [12.2.0] - 2019-09-27
### Changed
- Use aiohttp for autograph calls

## [12.1.0] - 2019-09-05
### Changed
- For authenticode signing, skip files that are already signed

## [12.0.0] - 2019-08-28
### Added
- Support for authenticode signing

## [11.1.0] - 2019-08-07
### Added
- Support for langpack signing

## [11.0.6] - 2019-07-29
### Added
- Added `.bin` to list of signable extensions
### Removed
- Obsolete SSL certificates for signing servers

## [11.0.5] - 2019-07-23
### Removed
- Datadog metric collection

## [11.0.4] - 2019-07-11
### Added
- New SSL certificates for signing servers
### Removed
- Obsolete SSL certificates for signing servers

## [11.0.3] - 2019-07-10
### Added
- New SSL certificates for signing servers

## [11.0.2] - 2019-06-12
### Fixed
- Fixed the `autograph_stage.pem` reference in `sign.py`

## [11.0.1] - 2019-06-10
### Fixed
- bugfix release to make sure vendored/ directory is properly included
  in whl and source distributions

## [11.0.0] - 2019-06-07
### Added
- added `vendored` directory for externally hosted code
- imported mozpack and mozbuild from mozilla-central at a specific rev supporting
  py3 in parts we use
- added support for omni.ja signing (as xpi signing) via autograph. Preserving
  gecko optimizations of the package.

## [10.1.0] - 2019-05-31
### Added
- added `log_level` kwarg to `execute_subprocess`

### Changed
- removed a lot of verbosity
- changed undmg log level to debug

## [10.0.0] - 2019-04-18
### Added
- added autograph_gpg and autograph_widevine signing functions
- Key ids can be specified by appending ':<keyid>' to the signing format. e.g.
  "autograph_hash_only_mar384:firefox_20190321_dev"

### Changed
- stop looking at signing scope format specifiers, and only pay attention to
  formats specified in upstreamArtifacts

## [9.6.0] - 2019-02-19
### Added
- `get_mar_verification_nick` and `verify_mar_signature` for mar signature verification.
- added `autograph_stage_mar384` format for autograph-stage testing.

### Changed
- `sign_mar384_with_autograph_hash` now verifies the mar signature.

### Removed
- removed `schema_file` from the example config.

## [9.5.1] - 2018-11-22
### Fixed
- APK SHA1 detection

## [9.5.0] - 2018-11-22
### Changed
- Renamed `autograph_fennec_sha1` into `autograph_apk_fennec_sha1`. It was unused.

### Added
- Any product matching `autograph_apk_*` will be signed as an APK.


## [9.4.1] - 2018-11-12
### Added
- Enable `autograph_fennec_sha1` as single signing format for Fennec APKs

## [9.4.0] - 2018-11-05
### Removed
- Strip existing APK signatures.

### Fixed
- Autograph doesn't recompress APKs anymore. This avoids start up crashes like https://github.com/mozilla-services/autograph/issues/164.

## [9.3.1] - 2018-10-29
### Fixed
- Strip existing APK signature before sending the APK to autograph. This maintains the behavior of the signing servers.

## [9.3.0] - 2018-10-23
### Added
- added msi signing support

## [9.2.0] - 2018-10-08
### Added
- Added new dep, `mar`. Python now requires lzma support; CentOS signing scriptworkers will need the xz-devel package.
- Added a new kwarg `raise_on_empty_list` to `get_suitable_signing_servers`. When True, and if the list of suitable signing servers is empty, the function raises a `SigningScriptError`.
- Added `sign_mar384_with_autograph_hash`. This is new functionality to only send the hash of the mar to autograph, to reduce autograph and network load.
- Added the `autograph_hash_only_mar384` signing format

### Changed
- Updated travis docker image to `mozilla/autograph:2.3.2`

## [9.1.0] - 2018-09-27
### Added
- Added optional `datadog_port` and `datadog_host` to the config file

### Changed
- Read the datadog config from the config file

## [9.0.3] - 2018-09-26
### Changed
- Updated signing server host cert to one with the new mac signing server alt names AND has a new subject (so it doesn't get ignored)

## [9.0.2] - 2018-09-25
### Changed
- Updated signing server host cert to one with the new mac signing server alt names

## [9.0.1] - 2018-09-17
### Fixed
- Initialize a hardcoded datadog statsd port of 8135 to avoid conflicting with collectd.

## [9.0.0] - 2018-09-17
### Changed
- Configuration: `taskcluster_scope_prefix` now becomes `taskcluster_scope_prefixes` and takes a JSON array.

## [8.0.1] - 2018-08-23
### Fixed
- fix typo in setup.py that caused the wrong package to be required

## [8.0.0] - 2018-08-22
### Added
- support for signing MARs and APKs with Autograph

## [7.0.4] - 2018-08-06
### Fixed
- catch `aiohttp.ClientError` and `asyncio.TimeoutError` during `get_token` requests

## [7.0.3] - 2018-07-27
### Changed
- remove the taskcluster pinning
- require py37 tests to be green

### Fixed
- create tarballs with root-owned files

## [7.0.2] - 2018-07-23
### Fixed
- Removed the old ssl cert, as this caused bustage.

## [7.0.1] - 2018-07-23
### Changed
- Updated the set of valid ssl certs

## [7.0.0] - 2018-05-10
### Added
- Added aiohttp3 support

### Changed
- Dropped py35 support; added py37 tests
- Packaged README.md

### Removed
- Removed README.rst

## [6.1.0] - 2018-04-24
### Added
- Added focus-jar support

## [6.0.1] - 2018-04-04
### Added
- Create `KEY` artifact when doing GPG signing

## [6.0.0] - 2018-03-14
### Changed
- `script.async_main()` relies on scriptworker (>= 10.2.0) to:
  - initialize context, config, and task
  - validate the task schema

### Removed
- `exceptions.TaskVerificationError` in favor of the one in scriptworker
- `script.SigningContext`, `script.usage()`, `script.main()` now handled by scriptworker
- `task.validate_task_schema()` now handled by scriptworker


## [5.0.0] - 2018-02-01
### Added
- support for different scope prefix (was hardcoded to `project:releng:signing:`). Prefixes are now defined in the configuration under `taskcluster_scope_prefix`.

tests/test_script.py
import mock
import os
import pytest

import scriptworker.client
from scriptworker.context import Context
from conftest import (
    noop_async,
    noop_sync,
    BASE_DIR,
)
import signingscript.script as script
from unittest.mock import MagicMock

# helper constants, fixtures, functions {{{1
EXAMPLE_CONFIG = os.path.join(BASE_DIR, "config_example.json")
SSL_CERT = os.path.join(BASE_DIR, "signingscript", "data", "host.cert")


# async_main {{{1
async def async_main_helper(
    tmpdir, mocker, formats, extra_config={}, server_type="signing_server"
):
    def fake_filelist_dict(*args, **kwargs):
        return {"path1": {"full_path": "full_path1", "formats": formats}}

    async def fake_sign(_, val, *args):
        return [val]

    mocker.patch.object(script, "load_signing_server_config", new=noop_sync)
    mocker.patch.object(script, "task_cert_type", new=noop_sync)
    mocker.patch.object(script, "task_signing_formats", return_value=formats)
    mocker.patch.object(script, "get_token", new=noop_async)
    mocker.patch.object(script, "build_filelist_dict", new=fake_filelist_dict)
    mocker.patch.object(script, "sign", new=fake_sign)
    context = mock.MagicMock()
    context.config = {"work_dir": tmpdir, "ssl_cert": None, "artifact_dir": tmpdir}
    context.config.update(extra_config)
    await script.async_main(context)


@pytest.mark.asyncio
async def test_async_main_gpg(tmpdir, tmpfile, mocker):
    formats = ["gpg"]
    fake_gpg_pubkey = tmpfile
    mocked_copy_to_dir = mocker.Mock()
    mocker.patch.object(script, "copy_to_dir", new=mocked_copy_to_dir)

    await async_main_helper(tmpdir, mocker, formats, {"gpg_pubkey": fake_gpg_pubkey})
    for call in mocked_copy_to_dir.call_args_list:
        if call[1].get("target") == "public/build/KEY":
            break
    else:
        assert False, "couldn't find copy_to_dir call that created KEY"


@pytest.mark.asyncio
async def test_async_main_gpg_no_pubkey_defined(tmpdir, mocker):
    formats = ["gpg"]

    try:
        await async_main_helper(tmpdir, mocker, formats)
    except Exception as e:
        assert e.args[0] == "GPG format is enabled but gpg_pubkey is not defined"


@pytest.mark.asyncio
async def test_async_main_gpg_pubkey_doesnt_exist(tmpdir, mocker):
    formats = ["gpg"]

    try:
        await async_main_helper(tmpdir, mocker, formats, {"gpg_pubkey": "faaaaaaake"})
    except Exception as e:
        assert e.args[0] == "gpg_pubkey (faaaaaaake) doesn't exist!"


@pytest.mark.asyncio
async def test_async_main_multiple_formats(tmpdir, mocker):
    formats = ["mar", "jar"]
    mocker.patch.object(script, "copy_to_dir", new=noop_sync)
    await async_main_helper(tmpdir, mocker, formats)


@pytest.mark.asyncio
async def test_async_main_autograph(tmpdir, mocker):
    formats = ["autograph_mar"]
    mocker.patch.object(script, "copy_to_dir", new=noop_sync)
    await async_main_helper(tmpdir, mocker, formats, {}, "autograph")


@pytest.mark.asyncio
async def test_craft_aiohttp_connector():
    context = Context()
    context.config = {}
    connector = script._craft_aiohttp_connector(context)
    assert connector._ssl is None

    context.config["ssl_cert"] = SSL_CERT
    connector = script._craft_aiohttp_connector(context)
    assert connector._ssl


def test_get_default_config():
    parent_dir = os.path.dirname(os.getcwd())
    c = script.get_default_config()
    assert c["work_dir"] == os.path.join(parent_dir, "work_dir")


def test_main(monkeypatch):
    sync_main_mock = MagicMock()
    monkeypatch.setattr(scriptworker.client, "sync_main", sync_main_mock)
    script.main()
    sync_main_mock.asset_called_once_with(
        script.async_main, default_config=script.get_default_config()
    )


@pytest.mark.asyncio
async def test_async_main_widevine_no_cert_defined(tmpdir, mocker):
    formats = ["autograph_widevine"]
    with pytest.raises(Exception) as e:
        await async_main_helper(tmpdir, mocker, formats)
        assert (
            e.args[0] == "Widevine format is enabled, but widevine_cert is not defined"
        )


@pytest.mark.asyncio
async def test_async_main_widevine(tmp_path, mocker):
    mocker.patch.object(script, "copy_to_dir")
    tmp_cert = tmp_path / "widevine.crt"
    formats = ["autograph_widevine"]
    await async_main_helper(tmp_path, mocker, formats, {"widevine_cert": tmp_cert})


docker.d/init.sh
#!/bin/bash
set -e

function check_var() {
    varname=$1
    if ! test "${!varname}"; then
        echo "${varname}" not set
        exit 1
    fi
}

#
# Check for certain variables which should be set
#
check_var PROJECT_NAME
check_var ENV
check_var COT_PRODUCT
check_var TASKCLUSTER_CLIENT_ID
check_var TASKCLUSTER_ACCESS_TOKEN
if [ "$ENV" == "prod" ]; then
  check_var ED25519_PRIVKEY
fi

#
# Validate content of certain variables
#
case $ENV in
  prod)
    export TRUST_LEVEL=3
    export WORKER_SUFFIX=
    ;;
  fake-prod)
    export TRUST_LEVEL=1
    # special case for signing, using -t- instead
    if [ $PROJECT_NAME = "signing" ]; then
        export TRUST_LEVEL=t
    fi
    export WORKER_SUFFIX=
    ;;
  dev)
    export TRUST_LEVEL=1
    if [ $PROJECT_NAME = "signing" ]; then
        export TRUST_LEVEL=t
    fi
    export WORKER_SUFFIX="-dev"
    ;;
  *)
    exit 1
    ;;
esac
case $COT_PRODUCT in
  firefox)
    export TRUST_DOMAIN=gecko
    ;;
  thunderbird)
    export TRUST_DOMAIN=comm
    ;;
  mobile)
    export TRUST_DOMAIN=mobile
    ;;
  application-services)
    export TRUST_DOMAIN=appservices
    ;;
  *)
    exit 1
    ;;
esac

#
# Defaults
#
# For development purposes uncomment the following line when pushing to dev branch
# export $COT_PRODUCT="thunderbird|mobile|..."
export ARTIFACTS_DIR=/app/artifacts
export ARTIFACT_UPLOAD_TIMEOUT=1200
export CONFIG_DIR=/app/configs
export CONFIG_LOADER=/app/bin/configloader
export ED25519_PRIVKEY_PATH=$CONFIG_DIR/ed25519_privkey
export GITHUB_OAUTH_TOKEN=
export LOGS_DIR=/app/logs
export PROVISIONER_ID=scriptworker-k8s
export SCRIPTWORKER=/app/bin/scriptworker
export SIGN_CHAIN_OF_TRUST=false
if [ "$ENV" == "prod" ]; then
  export SIGN_CHAIN_OF_TRUST=true
fi
export TASK_CONFIG=$CONFIG_DIR/worker.json
export TASK_LOGS_DIR=$ARTIFACTS_DIR/public/logs
export TASK_MAX_TIMEOUT=3600
export TASK_SCRIPT=/app/bin/${PROJECT_NAME}script
export TEMPLATE_DIR=/app/docker.d
export VERBOSE=true
export VERIFY_CHAIN_OF_TRUST=true
export VERIFY_COT_SIGNATURE=false
if [ "$ENV" == "prod" ]; then
  export VERIFY_COT_SIGNATURE=true
fi
export WORK_DIR=/app/workdir
export WORKER_TYPE="${TRUST_DOMAIN}-${TRUST_LEVEL}-${PROJECT_NAME}${WORKER_SUFFIX}"
export WORKER_GROUP=${WORKER_TYPE}
export WORKER_ID_PREFIX="${WORKER_TYPE}-"

#
# ensure configuration folder exists we can write to it
#
mkdir -p -m 700 $CONFIG_DIR
echo $ED25519_PRIVKEY | base64 -d > $ED25519_PRIVKEY_PATH
chmod 600 $ED25519_PRIVKEY_PATH

#
# run worker specific configuration
#
source $(dirname $0)/init_worker.sh

#
# create scriptworker and workerscript configuration
#
$CONFIG_LOADER --worker-id-prefix=$WORKER_ID_PREFIX $TEMPLATE_DIR/scriptworker.yml $CONFIG_DIR/scriptworker.json
$CONFIG_LOADER $TEMPLATE_DIR/worker.yml $CONFIG_DIR/worker.json


# unset all of the variables to not potentially leak them
exec env - /app/bin/scriptworker /app/configs/scriptworker.json


signingscript/task.py
#!/usr/bin/env python
"""Signingscript task functions.

Attributes:
    FORMAT_TO_SIGNING_FUNCTION (frozendict): a mapping between signing format
        and signing function. If not specified, use the `default` signing
        function.

"""
import aiohttp
import asyncio
from frozendict import frozendict
import logging
import os

from datadog import statsd, initialize
import platform

from scriptworker.exceptions import ScriptWorkerException, TaskVerificationError
from scriptworker.utils import retry_request, get_single_item_from_sequence

from signingscript.sign import get_suitable_signing_servers, sign_gpg, \
    sign_jar, sign_macapp, sign_signcode, sign_widevine, sign_file, \
    sign_mar384_with_autograph_hash
from signingscript.exceptions import SigningServerError
from signingscript.utils import is_autograph_signing_format

log = logging.getLogger(__name__)

FORMAT_TO_SIGNING_FUNCTION = frozendict({
    "autograph_hash_only_mar384": sign_mar384_with_autograph_hash,
    "gpg": sign_gpg,
    "jar": sign_jar,
    "focus-jar": sign_jar,
    "macapp": sign_macapp,
    "osslsigncode": sign_signcode,
    "sha2signcode": sign_signcode,
    # sha2signcodestub uses a generic sign_file
    "signcode": sign_signcode,
    "widevine": sign_widevine,
    "widevine_blessed": sign_widevine,
    "default": sign_file,
})


# task_cert_type {{{1
def task_cert_type(context):
    """Extract task certificate type.

    Args:
        context (Context): the signing context.

    Raises:
        TaskVerificationError: if the number of cert scopes is not 1.

    Returns:
        str: the cert type.

    """
    prefixes = _get_cert_prefixes(context)
    scopes = _extract_scopes_from_unique_prefix(
        scopes=context.task['scopes'],
        prefixes=prefixes
    )
    return get_single_item_from_sequence(
        scopes,
        condition=lambda _: True,     # scopes must just contain 1 single item
        ErrorClass=TaskVerificationError,
        no_item_error_message='No scope starting with any of these prefixes {} found'.format(prefixes),
        too_many_item_error_message='More than one scope found',
    )


# task_signing_formats {{{1
def task_signing_formats(context):
    """Get the list of signing formats from the task signing scopes.

    Args:
        context (Context): the signing context.

    Returns:
        list: the signing formats.

    """
    scopes = _extract_scopes_from_unique_prefix(
        scopes=context.task['scopes'],
        prefixes=_get_format_prefixes(context)
    )
    return [scope.split(':')[-1] for scope in scopes]


def _extract_scopes_from_unique_prefix(scopes, prefixes):
    scopes = [
         scope
         for scope in scopes
         for prefix in prefixes
         if scope.startswith(prefix)
    ]
    _check_scopes_exist_and_all_have_the_same_prefix(scopes, prefixes)
    return scopes


def _get_cert_prefixes(context):
    return _get_scope_prefixes(context, 'cert')


def _get_format_prefixes(context):
    return _get_scope_prefixes(context, 'format')


def _get_scope_prefixes(context, sub_namespace):
    prefixes = context.config['taskcluster_scope_prefixes']
    prefixes = [
        prefix if prefix.endswith(':') else '{}:'.format(prefix)
        for prefix in prefixes
    ]
    return ['{}{}:'.format(prefix, sub_namespace) for prefix in prefixes]


def _check_scopes_exist_and_all_have_the_same_prefix(scopes, prefixes):
    for prefix in prefixes:
        if all(scope.startswith(prefix) for scope in scopes):
            break
    else:
        raise TaskVerificationError(
            'Scopes must exist and all have the same prefix. '
            'Given scopes: {}. Allowed prefixes: {}'.format(scopes, prefixes)
        )


# get_token {{{1
async def get_token(context, output_file, cert_type, signing_formats):
    """Retrieve a token from the signingserver tied to my ip.

    Args:
        context (Context): the signing context
        output_file (str): the path to write the token to.
        cert_type (str): the cert type used to find an appropriate signing server
        signing_formats (list): the signing formats used to find an appropriate
            signing server

    Raises:
        SigningServerError: on failure

    """
    token = None
    data = {
        "slave_ip": context.config['my_ip'],
        "duration": context.config["token_duration_seconds"],
    }
    signing_servers = get_suitable_signing_servers(
        context.signing_servers, cert_type,
        [fmt for fmt in signing_formats if not is_autograph_signing_format(fmt)]
    )
    for s in signing_servers:
        log.info("getting token from %s", s.server)
        url = "https://{}/token".format(s.server)
        auth = aiohttp.BasicAuth(s.user, password=s.password)
        try:
            token = await retry_request(context, url, method='post', data=data,
                                        auth=auth, return_type='text')
            if token:
                break
        except (ScriptWorkerException, aiohttp.ClientError, asyncio.TimeoutError) as exc:
            log.warning("Error retrieving token: {}\nTrying the next server.".format(str(exc)))
            continue
    else:
        raise SigningServerError("Cannot retrieve signing token from any signing server.")
    with open(output_file, "w") as fh:
        print(token, file=fh, end="")


# sign {{{1
async def sign(context, path, signing_formats):
    """Call the appropriate signing function per format, for a single file.

    Args:
        context (Context): the signing context
        path (str): the source file to sign
        signing_formats (list): the formats to sign with

    Returns:
        list: the list of paths generated. This will be a list of one, unless
            there are detached sigfiles.

    """
    # We use 8135 by default, as datadog's default of 8125 conflicts
    # with collectd. Bug 1493265
    initialize(statsd_host=context.config.get('datadog_host', 'localhost'),
               statsd_port=context.config.get('datadog_port', 8135))

    output = path
    # Loop through the formats and sign one by one.
    for fmt in signing_formats:
        signing_func = FORMAT_TO_SIGNING_FUNCTION.get(
            fmt, FORMAT_TO_SIGNING_FUNCTION['default']
        )
        log.info("sign(): Signing {} with {}...".format(output, fmt))
        metric_tags = [
            'format:{}'.format(fmt),
            'host:{}'.format(platform.node()),
            'app:signingscript'
        ]
        with statsd.timed('signingfunc.time', tags=metric_tags):
            output = await signing_func(context, output, fmt)
    # We want to return a list
    if not isinstance(output, (tuple, list)):
        output = [output]
    return output


# _sort_formats {{{1
def _sort_formats(formats):
    """Order the signing formats.

    Certain formats need to happen before or after others, e.g. gpg after
    any format that modifies the binary.

    Args:
        formats (list): the formats to order.

    Returns:
        list: the ordered formats.

    """
    # Widevine formats must be after other formats other than macapp; GPG must
    # be last.
    for fmt in ("widevine", "widevine_blessed", "macapp", "gpg"):
        if fmt in formats:
            formats.remove(fmt)
            formats.append(fmt)
    return formats


# build_filelist_dict {{{1
def build_filelist_dict(context, all_signing_formats):
    """Build a dictionary of cot-downloaded paths and formats.

    Scriptworker will pre-download and pre-verify the `upstreamArtifacts`
    in our `work_dir`.  Let's build a dictionary of relative `path` to
    a dictionary of `full_path` and signing `formats`.

    Args:
        context (Context): the signing context
        all_signing_formats (list): the superset of valid signing formats,
            based on the task scopes.  If the file signing formats are not
            a subset, throw an exception.

    Raises:
        TaskVerificationError: if the files don't exist on disk, or the
            file signing formats are not a subset of all_signing_formats.

    Returns:
        dict of dicts: the dictionary of relative `path` to a dictionary with
            `full_path` and a list of signing `formats`.

    """
    filelist_dict = {}
    all_signing_formats_set = set(all_signing_formats)
    messages = []
    for artifact_dict in context.task['payload']['upstreamArtifacts']:
        for path in artifact_dict['paths']:
            full_path = os.path.join(
                context.config['work_dir'], 'cot', artifact_dict['taskId'],
                path
            )
            if not os.path.exists(full_path):
                messages.append("{} doesn't exist!".format(full_path))
            formats_set = set(artifact_dict['formats'])
            if not set(formats_set).issubset(all_signing_formats_set):
                messages.append("{} {} illegal format(s) {}!".format(
                    artifact_dict['taskId'], path,
                    formats_set.difference(all_signing_formats_set)
                ))
            filelist_dict[path] = {
                "full_path": full_path,
                "formats": _sort_formats(artifact_dict['formats']),
            }
    if messages:
        raise TaskVerificationError(messages)
    return filelist_dict
